api/annotate/annotate.go

```diff
@@ -0,0 +1,294 @@
+	crdInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			n := newObj.(*unstructured.Unstructured)
+			o := oldObj.(*unstructured.Unstructured)
+			newSpec := n.Object["spec"].(map[string]interface{})
+			oldSpec := o.Object["spec"].(map[string]interface{})
+			newStatus := n.Object["status"].(map[string]interface{})
+			oldStatus := o.Object["status"].(map[string]interface{})
+			// spec changes to indicate about log type changes and service name changes
+			if !api.DeepEqualMap(oldSpec, newSpec) {
+				specCh <- struct{}{} // Signal that the update occurred
+			}
+			// status changes to indicate about instrumentation status change
+			if !api.DeepEqualMap(oldStatus, newStatus) {
+				statusCh <- struct{}{}
+			}
+		},
+	})
+	// start watching for changes in crd
+	dynamicFactory.Start(ctx.Done())
+
+	// Create the response
+	response := ResourceAnnotateResponse{
+		Name:           resource.Name,
+		Namespace:      resource.Namespace,
+		ControllerKind: resource.ControllerKind,
+		LogType:        &resource.LogType,
+		ServiceName:    &resource.ServiceName,
+	}
+	response.ContainerName = resource.ContainerName
+	// choose the instrumentation annotation value and value according to the service name
+	actionValue := "true"
+	if resource.ServiceName == "" {
+		actionValue = "rollback"
+	}
+	// Update workload and custom resources
+	switch resource.ControllerKind {
+	case api.KindDeployment:
+		logger.Info("Updating deployment: ", resource.Name)
+		deployment, getErr := clientset.AppsV1().Deployments(resource.Namespace).Get(ctx, resource.Name, v1.GetOptions{})
+		if getErr != nil {
+			logger.Error(api.ErrorGet, getErr)
+			http.Error(w, api.ErrorGet+getErr.Error(), http.StatusInternalServerError)
+			return
+		}
+		// handle log type
+		if deployment.Spec.Template.ObjectMeta.Annotations == nil {
+			deployment.Spec.Template.ObjectMeta.Annotations = make(map[string]string)
+		}
+		if len(resource.LogType) != 0 {
+			deployment.Spec.Template.ObjectMeta.Annotations[LogTypeAnnotation] = resource.LogType
+		} else {
+			delete(deployment.Spec.Template.ObjectMeta.Annotations, LogTypeAnnotation)
+		}
+		// handle traces instrumentation annotations
+		// logz.io/instrument
+		deployment.Spec.Template.ObjectMeta.Annotations[InstrumentationAnnotation] = actionValue
+		// service name
+		if len(resource.ServiceName) != 0 {
+			deployment.Spec.Template.ObjectMeta.Annotations[ServiceNameAnnotation] = resource.ServiceName
+		} else {
+			delete(deployment.Spec.Template.ObjectMeta.Annotations, ServiceNameAnnotation)
+		}
+		_, err = clientset.AppsV1().Deployments(resource.Namespace).Update(ctx, deployment, v1.UpdateOptions{})
+		if err != nil {
+			logger.Error(api.ErrorUpdate, err)
+			http.Error(w, api.ErrorUpdate+err.Error(), http.StatusInternalServerError)
+			return
+		}
```

Comment:
If it's possible, I'd prefer it if for each case we will export the logic to a dedicated function (`handleUpdateDeployment`, `handleUpdateStatefulset` for example), since this function is already long.